org-code-javabuilder/lib/src/main/java/org/code/javabuilder/InputRedirectionStream.java

```diff
@@ -0,0 +1,98 @@
+package org.code.javabuilder;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.LinkedList;
+import java.util.Queue;
+
+/**
+ * An InputStream that queries an InputAdapter for new bytes. This is intended to redirect system.in
+ * to use the InputAdapter rather than the server's console. See
+ * https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html for full implementation
+ * details & contract.
+ */
+public class InputRedirectionStream extends InputStream {
+  private final Queue<Byte> queue;
+  private final InputAdapter inputAdapter;
+
+  public InputRedirectionStream(InputAdapter inputAdapter) {
+    this.inputAdapter = inputAdapter;
+    this.queue = new LinkedList<>();
+  }
+
+  /**
+   * See: https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#read-- Checks the queue
+   * for existing bytes. If the queue is empty, polls the inputAdapter for new data. This is a
+   * blocking call.
+   *
+   * @return the first byte in the queue
+   */
+  @Override
+  public int read() {
+    if (queue.peek() == null) {
+      byte[] message = inputAdapter.getNextMessage().getBytes(StandardCharsets.UTF_8);
+      for (byte b : message) {
+        queue.add(b);
+      }
+    }
+
+    return (int) queue.remove();
+  }
+
+  /** See: https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#read-byte:A- */
+  @Override
+  public int read(byte[] b) {
+    return read(b, 0, b.length);
+  }
+
+  /**
+   * See: https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#read-byte:A-int-int-
+   */
+  @Override
+  public int read(byte[] b, int off, int len) {
+    if (b == null) {
+      throw new NullPointerException();
+    }
+    if (off < 0 || len < 0 || len > b.length - off) {
+      throw new IndexOutOfBoundsException();
+    }
+
+    int k = 0;
+    while (k < len) {
+      b[off + k++] = (byte) read();
```

Comment:
nit: could we split `b[off + k++]` into a couple lines? At least one to increment `k`, maybe another to calculate the index we are inserting and give it a name, although `off + k` could be fine too.

Reply:
Good call. I'll do that.